pkg/ruler/ruler.go

```diff
@@ -518,6 +563,68 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
 	return ownsRuleGroup, nil
 }
 
+func (r *Ruler) LivenessCheck(_ context.Context, request *LivenessCheckRequest) (*LivenessCheckResponse, error) {
+	level.Debug(r.logger).Log("msg", "liveness check request", "request received from", request.RulerAddress)
+	if r.lifecycler.ServiceContext().Err() != nil || r.subservices.IsStopped() {
+		return nil, errors.New("ruler's context is canceled and might be stopping soon")
```

Comment:
I am unsure about this tbh. When `r.lifecycler.ServiceContext().Err() != nil || r.subservices.IsStopped()` condition is met, is the gRPC server still working correctly to handle liveness check?

Reply:
Depending upon how Ruler is configured, `r.lifecycler.ServiceContext()` can be canceled but gRPC server can be active for a long time. When `r.lifecycler.ServiceContext()` is canceled, Prometheus will stop evaluating rule groups as soon as possible but if the Ruler instance has not shutdown yet, it will respond to secondary as though it is active and this causes missed evaluations. This check is necessary.

`r.subservices.IsStopped()` is debatable. I only saw this in my testing in some cases during hard restarts (kill -9) when the gRPC server was already running but ring was not fully operational